collect_app/src/main/java/org/odk/collect/android/tasks/FormLoaderTask.java

```diff
@@ -270,22 +186,110 @@ protected FECWrapper doInBackground(String... path) {
         // Remove previous forms
         ReferenceManager.instance().clearSession();
 
+        processItemSets(formMediaDir);
+
+        // This should get moved to the Application Class
+        if (ReferenceManager.instance().getFactories().length == 0) {
+            // this is /sdcard/odk
+            ReferenceManager.instance().addReferenceFactory(new FileReferenceFactory(Collect.ODK_ROOT));
+        }
+
+        // Set jr://... to point to /sdcard/odk/forms/filename-media/
+        ReferenceManager.instance().addSessionRootTranslator(
+                new RootTranslator("jr://images/", "jr://file/forms/" + formFileName + "-media/"));
+        ReferenceManager.instance().addSessionRootTranslator(
+                new RootTranslator("jr://image/", "jr://file/forms/" + formFileName + "-media/"));
+        ReferenceManager.instance().addSessionRootTranslator(
+                new RootTranslator("jr://audio/", "jr://file/forms/" + formFileName + "-media/"));
+        ReferenceManager.instance().addSessionRootTranslator(
+                new RootTranslator("jr://video/", "jr://file/forms/" + formFileName + "-media/"));
+
+        final FormController fc = new FormController(formMediaDir, fec, instancePath == null ? null
+                : new File(instancePath));
+        if (xpath != null) {
+            // we are resuming after having terminated -- set index to this
+            // position...
+            FormIndex idx = fc.getIndexFromXPath(xpath);
+            fc.jumpToIndex(idx);
+        }
+        if (waitingXPath != null) {
+            FormIndex idx = fc.getIndexFromXPath(waitingXPath);
+            fc.setIndexWaitingForData(idx);
+        }
+        data = new FECWrapper(fc, usedSavepoint);
+        return data;
+    }
+
+    private FormDef createFormDefFromCacheOrXml(String formPath, File formXml) {
```

Comment:
Verified that this was copied correctly.

Reply:
It’s a shame that these diffs can’t follow an Extract Method refactoring.

Reply:
+1

Reply:
This.